internal/api/token_refresh.go

```diff
@@ -106,42 +106,60 @@ func (a *API) RefreshTokenGrant(ctx context.Context, w http.ResponseWriter, r *h
 			// refresh token row and session are locked at this
 			// point, cannot be concurrently refreshed
 
-			if token.Revoked {
-				a.clearCookieTokens(config, w)
-				// For a revoked refresh token to be reused, it
-				// has to fall within the reuse interval.
-
-				reuseUntil := token.UpdatedAt.Add(
-					time.Second * time.Duration(config.Security.RefreshTokenReuseInterval))
+			var issuedToken *models.RefreshToken
 
-				if time.Now().After(reuseUntil) {
-					// not OK to reuse this token
+			if token.Revoked {
+				activeRefreshToken, terr := session.FindCurrentlyActiveRefreshToken(tx)
+				if terr != nil {
+					return internalServerError(terr.Error())
+				}
 
-					if config.Security.RefreshTokenRotationEnabled {
-						// Revoke all tokens in token family
-						if err := models.RevokeTokenFamily(tx, token); err != nil {
-							return internalServerError(err.Error())
+				if activeRefreshToken != nil && activeRefreshToken.Parent.String() == token.Token {
+					// Token was revoked, but it's the
+					// parent of the currently active one.
+					// This indicates that the client was
+					// not able to store the result when it
+					// refreshed token. This case is
+					// allowed, provided we return back the
+					// active refresh token instead of
+					// creating a new one.
+					issuedToken = activeRefreshToken
+				} else {
+					// For a revoked refresh token to be reused, it
+					// has to fall within the reuse interval.
+					reuseUntil := token.UpdatedAt.Add(
```

Comment:
just wanted to check this `else {}` block here:
* If there is no `activeRefreshToken`, then it means that the token used in the request is the first token in the session right?
* If the `token` is revoked but it is in the reuse interval, do we want to allow swapping for a new refresh token? This would mean that if that token is being reused multiple times, it would generate multiple child refresh tokens - is this the behaviour we want?

Reply:
> If there is no activeRefreshToken, then it means that the token used in the request is the first token in the session right?

In theory, it means there's no active token in the session -- i.e. the user is "logged out." But because you can reuse a token again within some time period the session isn't really "gone." We should refactor this long term.

> If the token is revoked but it is in the reuse interval, do we want to allow swapping for a new refresh token? This would mean that if that token is being reused multiple times, it would generate multiple child refresh tokens - is this the behaviour we want?

Long term probably not. In general I think we can refactor the refresh tokens table to just store the revoked ones, and store the active one as a column on the sessions table. 

in any case, let's plan for changes like this gradually.